workspace/workspace_test.go

```diff
@@ -153,3 +153,117 @@ func TestLocate(t *testing.T) {
 		assert.Equal(t, test.out, dirs, test.desc)
 	}
 }
+
+func TestSolutionPath(t *testing.T) {
+	root := filepath.Join("..", "fixtures", "solution-path", "creatures")
+	ws := New(root)
+
+	// An existing exercise.
+	path, err := ws.SolutionPath("gazelle", "ccc")
+	assert.NoError(t, err)
+	assert.Equal(t, filepath.Join(root, "gazelle-3"), path)
+
+	path, err = ws.SolutionPath("gazelle", "abc")
+	assert.NoError(t, err)
+	assert.Equal(t, filepath.Join(root, "gazelle-4"), path)
+
+	// A new exercise.
+	path, err = ws.SolutionPath("lizard", "abc")
+	assert.NoError(t, err)
+	assert.Equal(t, filepath.Join(root, "lizard"), path)
+}
+
+func TestPathContainsSolution(t *testing.T) {
+	root := filepath.Join("..", "fixtures", "path-contains-solution")
+
+	ok, err := PathContainsSolution(filepath.Join(root, "yepp"), "abc")
+	assert.NoError(t, err)
+	assert.True(t, ok)
+
+	ok, err = PathContainsSolution(filepath.Join(root, "nope"), "abc")
+	assert.NoError(t, err)
+	assert.False(t, ok)
+
+	_, err = PathContainsSolution(filepath.Join(root, "broken"), "abc")
+	assert.Error(t, err)
+}
+
+func TestDetermineSolutionPath(t *testing.T) {
+	ws := New("tmp")
+
+	existsFn := func(path, solutionID string) (bool, error) {
+		pathToSolutionID := map[string]string{
+			filepath.Join(ws.Dir, "pig"):      "xxx",
+			filepath.Join(ws.Dir, "gecko"):    "aaa",
+			filepath.Join(ws.Dir, "gecko-2"):  "xxx",
+			filepath.Join(ws.Dir, "gecko-3"):  "ccc",
+			filepath.Join(ws.Dir, "bat"):      "aaa",
+			filepath.Join(ws.Dir, "dog"):      "aaa",
+			filepath.Join(ws.Dir, "dog-2"):    "bbb",
+			filepath.Join(ws.Dir, "dog-3"):    "ccc",
+			filepath.Join(ws.Dir, "rabbit"):   "aaa",
+			filepath.Join(ws.Dir, "rabbit-2"): "bbb",
+			filepath.Join(ws.Dir, "rabbit-4"): "ccc",
+		}
+		return pathToSolutionID[path] == solutionID, nil
+	}
```

Comment:
This is the mock function that lets us get away with not having fixture directories and fixture solution metadata files for all of these tests.